### PR TITLE
`azurerm_network_interface` - upgrade API from `2022-07-01` to `2023-02-01`

### PR DESCRIPTION
**Pull request by ms-zhenhua**
##  Description

Upgrade API version to support newly added properties: (new properties will be in another PR)
[auxiliaryMode](https://github.com/Azure/azure-rest-api-specs/blob/5758cc23b0022e403d876662d9799f02c9bba3e6/specification/network/resource-manager/Microsoft.Network/stable/2023-02-01/networkInterface.json#L1212)
[auxiliarySku](https://github.com/Azure/azure-rest-api-specs/blob/5758cc23b0022e403d876662d9799f02c9bba3e6/specification/network/resource-manager/Microsoft.Network/stable/2023-02-01/networkInterface.json#L1226)

## Test Result
image
image
image

--- PASS: TestAccNetworkInterface_disappears (309.71s)
--- PASS: TestAccNetworkInterface_ipv6 (388.27s)
--- PASS: TestAccNetworkInterface_multipleIPConfigurations (440.64s)
--- PASS: TestAccNetworkInterface_internalDomainNameLabel (480.79s)
--- PASS: TestAccNetworkInterface_pointToGatewayLB (534.90s)
--- PASS: TestAccNetworkInterface_multipleIPConfigurationsSecondaryAsPrimary (542.93s)
--- PASS: TestAccNetworkInterface_enableAcceleratedNetworking (546.67s)
--- PASS: TestAccNetworkInterface_enableIPForwarding (551.44s)
--- PASS: TestAccNetworkInterface_basic (570.84s)
--- PASS: TestAccNetworkInterface_dnsServers (582.01s)
--- PASS: TestAccNetworkInterface_tags (521.22s)
--- PASS: TestAccNetworkInterface_updateMultipleParameters (456.69s)
--- PASS: TestAccNetworkInterface_requiresImport (355.55s)
--- PASS: TestAccNetworkInterface_static (343.84s)
--- PASS: TestAccNetworkInterface_update (463.97s)
--- PASS: TestAccNetworkInterface_publicIP (391.04s)

--- PASS: TestAccDataSourceArmNetworkInterface_basic (227.24s)

--- PASS: TestAccNetworkInterfaceApplicationSecurityGroupAssociation_deleted (264.22s)
--- PASS: TestAccNetworkInterfaceApplicationSecurityGroupAssociation_multipleIPConfigurations (311.62s)
--- PASS: TestAccNetworkInterfaceApplicationSecurityGroupAssociation_requiresImport (338.72s)
--- PASS: TestAccNetworkInterfaceApplicationSecurityGroupAssociation_basic (369.77s)
--- PASS: TestAccNetworkInterfaceApplicationSecurityGroupAssociation_updateNIC (382.26s)

### CODE REVIEW

File: internal/services/network/network_interface_data_source.go [hunk 1]
```diff
@@ -7,10 +7,12 @@ import (
...
 	"time"
 
+	"github.com/hashicorp/go-azure-helpers/lang/response"
+	"github.com/hashicorp/go-azure-helpers/resourcemanager/commonids"
 	"github.com/hashicorp/go-azure-helpers/resourcemanager/commonschema"
+	"github.com/hashicorp/go-azure-sdk/resource-manager/network/2023-02-01/networkinterfaces"
 	"github.com/hashicorp/terraform-provider-azurerm/helpers/azure"
 	"github.com/hashicorp/terraform-provider-azurerm/internal/clients"
-	"github.com/hashicorp/terraform-provider-azurerm/internal/services/network/parse"
 	"github.com/hashicorp/terraform-provider-azurerm/internal/tags"
```
Comment (mbfrahry) [inline internal/services/network/network_interface_data_source.go:16]: This import wants to be swapped to 

```suggestion
	"github.com/hashicorp/go-azure-helpers/resourcemanager/tags"
```

And the schema will be

```
"tags": commonschema.TagsDataSource(),
```

File: internal/services/network/network_interface_data_source.go [hunk 2]
```diff
@@ -264,5 +272,9 @@ func dataSourceNetworkInterfaceRead(d *pluginsdk.ResourceData, meta interface{})
 		d.Set("enable_accelerated_networking", props.EnableAcceleratedNetworking)
 	}
 
-	return tags.FlattenAndSet(d, resp.Tags)
```
Comment (mbfrahry) [inline internal/services/network/network_interface_data_source.go:267]: Doing that will let us keep this line the same without doing what you've done below

File: internal/services/network/network_interface_resource.go [hunk 3]
```diff
@@ -364,142 +363,146 @@ func resourceNetworkInterfaceUpdate(d *pluginsdk.ResourceData, meta interface{})
...
+		update.Properties.IPConfigurations = ipConfigs
 	} else {
-		update.InterfacePropertiesFormat.IPConfigurations = existing.InterfacePropertiesFormat.IPConfigurations
+		update.Properties.IPConfigurations = existing.Model.Properties.IPConfigurations
 	}
 
 	if d.HasChange("tags") {
 		tagsRaw := d.Get("tags").(map[string]interface{})
-		update.Tags = tags.Expand(tagsRaw)
+		update.Tags = utils.ExpandPtrMapStringString(tagsRaw)
```
Comment (mbfrahry) [inline internal/services/network/network_interface_resource.go:373]: Apply that same `tag` logic from the datasource here

**Review by mbfrahry — DISMISSED** [review]
Hey @ms-zhenhua, this is mostly good but `tags` isn't quite right and I've detailed how to fix it going forward for future sdk swaps

File: internal/services/network/network_interface.go [hunk 4]
```diff
@@ -87,50 +87,54 @@ func parseFieldsFromNetworkInterface(input network.InterfacePropertiesFormat) ne
...
-		if config.InterfaceIPConfigurationPropertiesFormat.PrivateIPAddressVersion != network.IPVersionIPv4 {
+		if config.Properties.PrivateIPAddressVersion == nil {
 			continue
 		}
 
-		config.ApplicationSecurityGroups = &applicationSecurityGroups
-		config.ApplicationGatewayBackendAddressPools = &applicationGatewayBackendAddressPools
-		config.LoadBalancerBackendAddressPools = &loadBalancerBackendAddressPools
-		config.LoadBalancerInboundNatRules = &loadBalancerInboundNatRules
+		if *config.Properties.PrivateIPAddressVersion != networkinterfaces.IPVersionIPvFour {
```
[the remark this answers is not in the document]
Comment (ms-zhenhua) [reply internal/services/network/network_interface.go]: Code updated.

File: internal/services/network/network_interface_data_source.go (hunk 1 above)
Comment (ms-zhenhua) [reply internal/services/network/network_interface_data_source.go]: Code updated.

File: internal/services/network/network_interface_resource.go (hunk 3 above)
Comment (ms-zhenhua) [reply internal/services/network/network_interface_resource.go]: Code updated.

**Comment by ms-zhenhua** [comment]
Hi @mbfrahry, thank you for your review. I have updated the code. Kindly take another review.

File: internal/services/network/edge_zone.go [hunk 5]
```diff
@@ -27,3 +27,21 @@ func flattenEdgeZone(input *network.ExtendedLocation) string {
...
+	return &edgezones.Model{
+		Name: normalized,
+	}
+}
+
+func flattenEdgeZoneModel(input *edgezones.Model) string {
+	if input == nil || input.Name == "" {
+		return ""
+	}
+	return edgezones.NormalizeNilable(&input.Name)
```
Comment (tombuildsstuff) [inline internal/services/network/edge_zone.go:46]: we don't need to pass this as a pointer, we can instead use the regular `Normalize` method

```suggestion
	return edgezones.Normalize(input.Name)
```

**Review by tombuildsstuff — APPROVED** [review]
hey @ms-zhenhua 

Thanks for pushing those changes - I've taken a look and left one comment inline, since this needs a rebase and we're going to want to move the `Network` package to using a Meta Client sooner rather than later, I hope you don't mind but I'm going to push a commit to fix this comment/rebase this so that we can get this merged.

Thanks!

**Pull request merged**